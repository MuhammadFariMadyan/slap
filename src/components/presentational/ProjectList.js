import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import "typeface-roboto";

import Project from "./Project";

const Container = styled.div`
  display: grid;
  grid-template-columns: repeat(4, 240px);
  grid-template-rows: auto;
  flex-wrap: wrap;
  width: 100%;
  justify-content: space-between;
`;

const Empty = styled.p`
  font-size: 14px;
  font-family: Roboto;
  color: #333;
  margin: 0px;
`;

const ProjectList = ({ className, items, onClick, onClickDelete }) => (
  <Container className={className}>
    {items.length ? (
      items.map(item => (
        <Project
          key={item.id}
          {...item}
          onClick={() => onClick(item)}
          onClickDelete={() => onClickDelete(item)}
        />
      ))
    ) : (
      <Empty>No projects found.</Empty>
    )}
  </Container>
);

ProjectList.propTypes = {
  className: PropTypes.string,
  items: PropTypes.array,
  onClick: PropTypes.func,
  onClickDelete: PropTypes.func
};

ProjectList.defaultProps = {
  items: []
};

export default ProjectList;
