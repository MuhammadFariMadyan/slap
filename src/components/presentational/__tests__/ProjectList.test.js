import React from "react";
import { shallow } from "enzyme";
import "jest-styled-components";

import ProjectList from "../ProjectList";

describe("ProjectList", () => {
  jest.mock("../Project", () => "Project");

  const setup = propOverrides => {
    const props = {
      items: [...Array(3)].map((_, index) => ({
        id: `project-${index + 1}`,
        name: `Project ${index + 1}`,
        timestamp: 1537004755445,
        contents: [
          {
            id: "content-1",
            slideID: "slide-1",
            projectID: "project-1",
            type: "text",
            content: "Hello World",
            position: {
              x: 32,
              y: 32
            }
          },
          {
            id: "content-2",
            slideID: "slide-1",
            projectID: "project-1",
            type: "text",
            content: "Stay hungry",
            position: {
              x: 128,
              y: 96
            }
          }
        ]
      })),
      onClick: jest.fn(),
      onClickDelete: jest.fn(),
      ...propOverrides
    };

    const wrapper = shallow(<ProjectList {...props} />);

    return {
      wrapper,
      props
    };
  };

  it("should renders default", () => {
    const { wrapper } = setup();

    expect(wrapper).toMatchSnapshot();
  });

  it("should renders without items", () => {
    const { wrapper } = setup({ items: [] });

    expect(wrapper).toMatchSnapshot();
  });

  it("should calls 'onClick' callback while name clicked", () => {
    const { wrapper, props } = setup();

    props.items.forEach((item, index) => {
      wrapper
        .find("Project")
        .at(index)
        .props()
        .onClick(item);

      expect(props.onClick).toBeCalledWith(item);
    });
  });

  it("should calls 'onClickDelete' callback while delete clicked", () => {
    const { wrapper, props } = setup();

    props.items.forEach((item, index) => {
      wrapper
        .find("Project")
        .at(index)
        .props()
        .onClickDelete(item);

      expect(props.onClickDelete).toBeCalledWith(item);
    });
  });
});
