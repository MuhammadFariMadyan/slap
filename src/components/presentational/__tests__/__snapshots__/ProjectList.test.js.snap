// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProjectList should renders default 1`] = `
<styled.div>
  <Project
    contents={
      Array [
        Object {
          "content": "Hello World",
          "id": "content-1",
          "position": Object {
            "x": 32,
            "y": 32,
          },
          "projectID": "project-1",
          "slideID": "slide-1",
          "type": "text",
        },
        Object {
          "content": "Stay hungry",
          "id": "content-2",
          "position": Object {
            "x": 128,
            "y": 96,
          },
          "projectID": "project-1",
          "slideID": "slide-1",
          "type": "text",
        },
      ]
    }
    id="project-1"
    key="project-1"
    name="Project 1"
    onClick={[Function]}
    onClickDelete={[Function]}
    timestamp={1537004755445}
  />
  <Project
    contents={
      Array [
        Object {
          "content": "Hello World",
          "id": "content-1",
          "position": Object {
            "x": 32,
            "y": 32,
          },
          "projectID": "project-1",
          "slideID": "slide-1",
          "type": "text",
        },
        Object {
          "content": "Stay hungry",
          "id": "content-2",
          "position": Object {
            "x": 128,
            "y": 96,
          },
          "projectID": "project-1",
          "slideID": "slide-1",
          "type": "text",
        },
      ]
    }
    id="project-2"
    key="project-2"
    name="Project 2"
    onClick={[Function]}
    onClickDelete={[Function]}
    timestamp={1537004755445}
  />
  <Project
    contents={
      Array [
        Object {
          "content": "Hello World",
          "id": "content-1",
          "position": Object {
            "x": 32,
            "y": 32,
          },
          "projectID": "project-1",
          "slideID": "slide-1",
          "type": "text",
        },
        Object {
          "content": "Stay hungry",
          "id": "content-2",
          "position": Object {
            "x": 128,
            "y": 96,
          },
          "projectID": "project-1",
          "slideID": "slide-1",
          "type": "text",
        },
      ]
    }
    id="project-3"
    key="project-3"
    name="Project 3"
    onClick={[Function]}
    onClickDelete={[Function]}
    timestamp={1537004755445}
  />
</styled.div>
`;

exports[`ProjectList should renders without items 1`] = `
<styled.div>
  <styled.p>
    No projects found.
  </styled.p>
</styled.div>
`;
